crates/papyrus_base_layer/src/eth_events.rs

```diff
@@ -23,8 +26,27 @@ impl TryFrom<Log> for L1Event {
 
         let event = Starknet::StarknetEvents::decode_log(&log, validate)?.data;
         match event {
-            Starknet::StarknetEvents::LogMessageToL2(_event) => {
-                todo!()
+            Starknet::StarknetEvents::LogMessageToL2(event) => {
+                let fee =
+                    Fee(event.fee.try_into().map_err(EthereumBaseLayerError::FeeOutOfRange)?);
+                let event_data = EventData::try_from(event)?;
+                // Prepend the L1 sender address to the calldata.
+                let calldata = Calldata(Arc::new(
+                    once(event_data.from_address.into())
+                        .chain(Arc::try_unwrap(event_data.payload.0).unwrap())
```

Comment:
Would this result in us cloning each piece of the vector instead of moving the content of `payload`?
<!-- Reviewable comment -OEwttVT3dxuRcNLgH_K:b-e8k2lb -->
<!-- Sent from Reviewable.io -->

Reply:
Nope, clone is never implicit, everything here is consuming.
<!-- Reviewable comment -OExXufDBsLrmSpJuhxe:bi0vc47 -->
<!-- Sent from Reviewable.io -->